Remove External Workload

brb opened the pull request:
Needs clarification (:wave:  @giorio94):

- [x] Is `option.Config.JoinCluster` used only by the external-workloads feature? Discovered via https://github.com/cilium/cilium/blob/pr/brb/rm-ext-wl/pkg/health/server/server.go#L446.
- [x] Shall we change the client-secret name - https://github.com/cilium/cilium/blob/pr/brb/rm-ext-wl/install/kubernetes/cilium/templates/clustermesh-apiserver/tls-helm/client-secret.yaml#L3? Or can we remove it? From the top if-statement it looks that it is used by others too.
- [x] Ditto - https://github.com/cilium/cilium/blob/pr/brb/rm-ext-wl/install/kubernetes/cilium/templates/clustermesh-apiserver/tls-certmanager/client-secret.yaml#L20?

```release-note
Remove external-workload feature from Cilium and cilium-cli. Use previous version of cilium-cli if needed to provision unsupported external-workload clusters.
```

brb committed:
Remove External Workloads feature

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

brb committed:
clustermesh: Make id/name validation strict

0/"default" were valid only for CEW. Since the entire feature was
removed in the previous commit, we can validate for the != 0/"default".

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

giorio94 commented on pkg/clustermesh/types/types.go:
```diff
@@ -77,6 +80,20 @@ func ValidateClusterName(name string) error {
 	return nil
 }
 
+func RegisterClusterInfoValidator(lc cell.Lifecycle, cinfo ClusterInfo, log logrus.FieldLogger) {
```
Nit. the `log` parameter is unused, and can be dropped.

giorio94 commented on cilium-cli/cli/clustermesh.go:
We cannot unfortunately drop external workloads from the Cilium CLI yet, because the feature still exists in stable branches, and removing it would break CI there. Similarly, I think we'll need to keep the CEW type (and corresponding k8s client) definitions, as used by the CLI.

Ideally, we should add a version check to the CLI in order to return an entry in case one of the commands is run against Cilium v1.18 or later. Additionally the commands should be probably marked as deprecated to make it clear that they are no longer supported in combination with newer versions of Cilium.

brb replied:
As discussed with Joe Stringer offline, I will remove the EW CI workflows from the stable branches.

giorio94 commented on clustermesh-apiserver/clustermesh/root.go:
```diff
@@ -385,7 +383,7 @@ func startServer(
 		Endpoints:    resources.Endpoints,
 		Backend:      backend,
-		SharedOnly:   !allServices,
+		SharedOnly:   false,
```
`SharedOnly` should be set to `true`. Even better, the parameter could be dropped altogether, as all callers would set it to true now.

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-apiserver/tls-helm/client-secret.yaml:
This file can be dropped, as only used for external workloads.

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-apiserver/tls-certmanager/client-secret.yaml:
This file can be dropped, as only used for external workloads.

giorio94 reviewed (changes requested):
Thanks of taking care of this cleanup!

A couple of comments inline, mostly around the need of dropping the CLI changes to not break stable branches. 

Back to your questions:

>  Is option.Config.JoinCluster used only by the external-workloads feature? Discovered via https://github.com/cilium/cilium/blob/pr/brb/rm-ext-wl/pkg/health/server/server.go#L446.

Yes, that is the agent-side implementation of external workloads. The flag can be dropped, as well as all the code blocks guarded by it.

> Shall we change the client-secret name - https://github.com/cilium/cilium/blob/pr/brb/rm-ext-wl/install/kubernetes/cilium/templates/clustermesh-apiserver/tls-helm/client-secret.yaml#L3? Or can we remove it? From the top if-statement it looks that it is used by others too.

That file can be dropped, as only required by external workloads.

> Ditto - https://github.com/cilium/cilium/blob/pr/brb/rm-ext-wl/install/kubernetes/cilium/templates/clustermesh-apiserver/tls-certmanager/client-secret.yaml#L20?

Same here.

brb committed:
daemon: Remove JoinCluster

According to Marco Iorio, the JoinCluster was used only by the External
Workloads feature.

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

giorio94 commented on install/kubernetes/cilium/templates/validate.yaml:
```diff
@@ -169,14 +169,6 @@
     {{ fail "The clustermesh-apiserver cannot be enabled in combination with .Values.disableEndpointCRD=true" }}
   {{- end }}
 {{- end }}
-{{- if .Values.externalWorkloads.enabled }}
```
Nit. There is one more occurrence to delete at line 182 (introduced by another PR recently merged).

giorio94 commented on pkg/clustermesh/types/types.go:
Nit. These changes should be moved to the second commit.

giorio94 commented on daemon/cmd/daemon.go:
```diff
@@ -721,39 +718,6 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
-
-		// This can override node addressing config, so do this before starting IPAM
-		log.WithField(logfields.NodeName, nodeTypes.GetName()).Debug("Calling JoinCluster()")
-		if err := d.nodeDiscovery.JoinCluster(nodeTypes.GetName()); err != nil {
```
If you are up for some more clean-up, this function could be deleted as well (and in turn a few more dependents). Otherwise I'll take care of that in a follow-up PR.

brb replied:
:+1: Added one more commit.

giorio94 reviewed (approved):
Thanks! :broom:

brb committed:
nodediscovery: Remove JoinCluster

No longer used.

Suggested-by: Marco Iorio <marco.iorio@isovalent.com>
Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

pippolo84 reviewed (approved):
Operator changes LGTM :heavy_check_mark:

derailed reviewed (approved):
@brb Nice work! Looks like there are conflicts...
